test/framework/src/main/java/org/elasticsearch/test/InternalTestCluster.java

```diff
@@ -583,8 +613,27 @@ private NodeAndClient buildNode(int nodeId, long seed, Settings settings, Versio
         return new NodeAndClient(name, node);
     }
 
-    private String buildNodeName(int id) {
-        return nodePrefix + id;
+    private String buildNodeName(int id, Settings settings) {
+        String prefix = nodePrefix;
+        prefix = prefix + getRoleSuffix(settings);
+        return prefix + id;
+    }
+
+    /** returns a suffix string based on the node role. If no explicit role is defined, the suffix will be empty */
+    private String getRoleSuffix(Settings settings) {
+        String suffix = "";
+        if (Node.NODE_MASTER_SETTING.exists(settings) && Node.NODE_MASTER_SETTING.get(settings)) {
+            suffix = suffix + "m";
```

Comment:
+1 on using the relevant role's getAbbreviation() - I didn't realize its there (or it wasn't when I wrote this).  I didn't use all the nodes roles because  I didn't feel that ingest matters here. Also, I opted to not have a suffix at all if everything is default - that's something we can choose to do differently though.